File: articles/ethics-power-inclusion-design-thinking/go/ethical_risk_engine.go
package main

import (
	"encoding/csv"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
)

type DesignDecision struct {
	Name           string
	Harm           float64
	Probability    float64
	Exposure       float64
	Detectability  float64
	Accountability float64
	Inclusion      float64
	PublicValue    float64
	Repairability  float64
	Privacy        float64
	Autonomy       float64
	Manipulation   float64
	EthicalRisk    float64
	ReviewPriority float64
}

func parseFloat(value string) float64 {
	parsed, err := strconv.ParseFloat(value, 64)
	if err != nil {
		log.Fatalf("unable to parse numeric value %q: %v", value, err)
	}
	return parsed
}

func ethicalRisk(x DesignDecision) float64 {
	return x.Harm * x.Probability * x.Exposure * (1.0 - x.Detectability) * (1.0 - x.Accountability)
}

func reviewPriority(x DesignDecision) float64 {
	repairDeficit := 1.0 - x.Repairability
	return 0.32*x.EthicalRisk +
		0.20*x.Harm +
		0.14*x.Exposure +
		0.10*(1.0-x.Accountability) +
		0.08*(1.0-x.Detectability) +
		0.06*(1.0-x.Inclusion) +
		0.05*x.Privacy +
		0.03*x.Autonomy +
		0.02*x.Manipulation -
		0.12*x.PublicValue +
		0.10*repairDeficit
}

func loadDecisions(path string) []DesignDecision {
	file, err := os.Open(path)
	if err != nil {
		log.Fatalf("unable to open input file: %v", err)
	}
	defer file.Close()

	reader := csv.NewReader(file)
	rows, err := reader.ReadAll()
	if err != nil {
		log.Fatalf("unable to read CSV: %v", err)
	}

	var items []DesignDecision

	for i, row := range rows {
		if i == 0 {
			continue
		}
		if len(row) < 13 {
			log.Fatalf("malformed row: %v", row)
		}

		x := DesignDecision{
			Name:           row[0],
			Harm:           parseFloat(row[2]),
			Probability:    parseFloat(row[3]),
			Exposure:       parseFloat(row[4]),
			Detectability:  parseFloat(row[5]),
			Accountability: parseFloat(row[6]),
			Inclusion:      parseFloat(row[7]),
			PublicValue:    parseFloat(row[8]),
			Repairability:  parseFloat(row[9]),
			Privacy:        parseFloat(row[10]),
			Autonomy:       parseFloat(row[11]),
			Manipulation:   parseFloat(row[12]),
		}
		x.EthicalRisk = ethicalRisk(x)
		x.ReviewPriority = reviewPriority(x)
		items = append(items, x)
	}

	return items
}

func main() {
	path := "../data/raw/ethical_design_decisions_raw.csv"
	if len(os.Args) > 1 {
		path = os.Args[1]
	}

	items := loadDecisions(path)

	sort.Slice(items, func(i, j int) bool {
		return items[i].ReviewPriority > items[j].ReviewPriority
	})

	fmt.Println("rank,design_decision,ethical_risk,review_priority")
	for i, x := range items {
		fmt.Printf("%d,%s,%.6f,%.6f\n", i+1, x.Name, x.EthicalRisk, x.ReviewPriority)
	}
}

File: articles/ethics-power-inclusion-design-thinking/cpp/ethical_risk_engine.cpp
#include <algorithm>
#include <fstream>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <stdexcept>
#include <string>
#include <vector>

struct DesignDecision {
    std::string name;
    double harm;
    double probability;
    double exposure;
    double detectability;
    double accountability;
    double inclusion;
    double public_value;
    double repairability;
    double privacy;
    double autonomy;
    double manipulation;
};

struct ScoredDecision {
    std::string name;
    double ethical_risk;
    double review_priority;
};

std::vector<std::string> split_csv_line(const std::string& line) {
    std::vector<std::string> fields;
    std::stringstream ss(line);
    std::string item;
    while (std::getline(ss, item, ',')) {
        fields.push_back(item);
    }
    return fields;
}

double to_double(const std::string& value) {
    try {
        return std::stod(value);
    } catch (...) {
        throw std::runtime_error("Unable to parse numeric value: " + value);
    }
}

double ethical_risk(const DesignDecision& x) {
    return x.harm * x.probability * x.exposure * (1.0 - x.detectability) * (1.0 - x.accountability);
}

double review_priority(const DesignDecision& x) {
    const double r = ethical_risk(x);
    const double repair_deficit = 1.0 - x.repairability;
    return 0.32 * r +
           0.20 * x.harm +
           0.14 * x.exposure +
           0.10 * (1.0 - x.accountability) +
           0.08 * (1.0 - x.detectability) +
           0.06 * (1.0 - x.inclusion) +
           0.05 * x.privacy +
           0.03 * x.autonomy +
           0.02 * x.manipulation -
           0.12 * x.public_value +
           0.10 * repair_deficit;
}

std::vector<DesignDecision> read_decisions(const std::string& path) {
    std::ifstream file(path);
    if (!file) {
        throw std::runtime_error("Unable to open input file: " + path);
    }

    std::vector<DesignDecision> decisions;
    std::string line;
    std::getline(file, line);

    while (std::getline(file, line)) {
        if (line.empty()) {
            continue;
        }

        auto fields = split_csv_line(line);
        if (fields.size() < 13) {
            throw std::runtime_error("Malformed CSV row: " + line);
        }

        decisions.push_back(
            DesignDecision{
                fields[0],
                to_double(fields[2]),
                to_double(fields[3]),
                to_double(fields[4]),
                to_double(fields[5]),
                to_double(fields[6]),
                to_double(fields[7]),
                to_double(fields[8]),
                to_double(fields[9]),
                to_double(fields[10]),
                to_double(fields[11]),
                to_double(fields[12])
            }
        );
    }

    return decisions;
}

int main(int argc, char* argv[]) {
    const std::string default_path = "../data/raw/ethical_design_decisions_raw.csv";
    const std::string input_path = argc > 1 ? argv[1] : default_path;

    try {
        auto decisions = read_decisions(input_path);
        std::vector<ScoredDecision> scored;

        for (const auto& decision : decisions) {
            scored.push_back(
                ScoredDecision{
                    decision.name,
                    ethical_risk(decision),
                    review_priority(decision)
                }
            );
        }

        std::sort(
            scored.begin(),
            scored.end(),
            [](const ScoredDecision& a, const ScoredDecision& b) {
                return a.review_priority > b.review_priority;
            }
        );

        std::cout << "rank,design_decision,ethical_risk,review_priority\n";
        for (std::size_t i = 0; i < scored.size(); ++i) {
            std::cout << (i + 1) << ","
                      << scored[i].name << ","
                      << std::fixed << std::setprecision(6)
                      << scored[i].ethical_risk << ","
                      << scored[i].review_priority << "\n";
        }
    } catch (const std::exception& ex) {
        std::cerr << "ERROR: " << ex.what() << "\n";
        return 1;
    }

    return 0;
}
